qcodes/instrument_drivers/QDevil/QDevil_QDAC.py

```diff
@@ -0,0 +1,938 @@
+        # function as opposed to when snapshot is called on an individual
+        # channel
+        snap = super().snapshot_base(
+                                update=update,
+                                params_to_skip_update=params_to_skip_update)
+        self._get_status_performed = False
+        return snap
+
+    #########################
+    # Channel gets/sets
+    #########################
+
+    def _set_voltage(self, chan, v_set):
+        """
+        set_cmd for the chXX_v parameter
+
+        Args:
+            chan (int): The 1-indexed channel number
+            v_set (float): The target voltage
+
+        If a finite slope has been assigned, a function generator will
+        ramp the voltage.
+        """
+
+        slope = self._slopes.get(chan, None)
+        if slope:
+            # We need .get and not cache/get_latest in case a ramp
+            # was interrupted
+            v_start = self.channels[chan-1].v.get()
+            duration = abs(v_set-v_start)/slope
+            LOG.info('Slope: {}, time: {}'.format(slope, duration))
+            # SYNCing happens inside ramp_voltages
+            self.ramp_voltages([chan], [v_start], [v_set], duration)
+        else:
+            self.write('wav {ch} 0 0 0;set {ch} {voltage:.6f}'
+                       .format(ch=chan, voltage=v_set))
+
+    # Helper for _set_mode. It is not possible ot say if the channel is
+    # connected to a generator, so we need to ask.
+    def _wav_or_set_msg(self, chan, new_voltage):
+        self.write('wav {}'.format(chan))
+        fw_str = self._write_response
+        gen, _, _ = fw_str.split(',')
+        if int(gen) > 0:
+            # The amplitude must be set to zero to avoid potential overflow.
+            # Assuming that voltage range is not changed during a ramp
+            return 'wav {} {} {:.6f} {:.6f}'\
+                    .format(chan, int(gen), 0, new_voltage)
+        else:
+            return 'set {} {:.6f}'.format(chan, new_voltage)
+
+    def _set_mode(self, chan, new_mode):
+        """
+        set_cmd for the QDAC's mode (combined voltage and current sense range)
+        """
+        # Mode 0: (10V, 100uA), Mode 1: (10V, 1uA), Mode 2: (1V, 1uA)
+        ivrangedict = {
+                        0: {"vol": 0, "cur": 1},
+                        1: {"vol": 0, "cur": 0},
+                        2: {"vol": 1, "cur": 0}}
+        # switchint =  (0,1,2,
+        #               3,4,5,
+        #               6,7,8)
+
+        old_mode = self.channels[chan-1].mode.cache()
+        new_vrange = ivrangedict[new_mode]["vol"]
+        new_irange = ivrangedict[new_mode]["cur"]
+        switchint = int(3*old_mode+new_mode)
+        message = ''
+
+        if old_mode == new_mode:
+            return
+
+        # If the voltage range is going to change we have to take care of
+        # setting the voltage after the switch, and therefore read them first
+        # Only the current range has to change
```

Comment:
I think this sounds very dangerous: From the comment it sounds that if the voltage range is changed from attenuation to no-attenuation, you will get a voltage spike.

Reply:
Partly addressed with the mode_force parameter. But at QDevil we prefer that the voltage stays when the range is changed, even though a spike might occur. Anyway, the new "mode_force" behaviour now takes care of protecting the user from spikes.